server/src/main/java/org/opensearch/common/cache/tier/OpenSearchOnHeapCache.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.cache.tier;
+
+import org.opensearch.common.cache.Cache;
+import org.opensearch.common.cache.CacheBuilder;
+import org.opensearch.common.cache.RemovalListener;
+import org.opensearch.common.cache.RemovalNotification;
+import org.opensearch.common.unit.TimeValue;
+
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.function.ToLongBiFunction;
+
+/**
+ * This variant of on-heap cache uses OpenSearch custom cache implementation.
+ * @param <K> Type of key
+ * @param <V> Type of value
+ */
+public class OpenSearchOnHeapCache<K, V> implements OnHeapCachingTier<K, V>, RemovalListener<K, V> {
+
+    private final Cache<K, V> cache;
```

Comment:
While I am generally in favor of composition, just wondering if it is possible to extend Cache instead of composing it? If yes, any downsides of doing that?

Reply:
By using inheritance, we won't be able to plug-in any other external implementation of onHeap cache like Guava or caffeine.

We can extend this Cache with OnHeapCachingTier as this Cache is built-in cache within OpenSearch where we have control. For any external libraries, we won't be able to do it and will only have to rely on composition like this.

Let me know if it make sense.

Reply:
Yes, that makes sense. Thank you!